client.go

```diff
@@ -326,28 +457,169 @@ func (c *Client) send(call *Call) {
 
 // receiveResponse reads a response to an RPC call
 func receiveResponse(s *streamWrap, call *Call) error {
-	logger.Debugf(
-		"waiting response for %s.%s to %s",
-		call.SvcID.Name,
-		call.SvcID.Method,
-		call.Dest,
-	)
+	logger.Debugf("waiting response for %s to %s", call.SvcID, call.Dest)
 	var resp Response
 	if err := s.dec.Decode(&resp); err != nil {
 		call.doneWithError(newClientError(err))
 		return err
 	}
 
-	defer call.done()
 	if e := resp.Error; e != "" {
-		call.setError(responseError(resp.ErrType, e))
+		err := responseError(resp.ErrType, e)
+		call.setError(err)
+		// we still try to read the body if possible
 	}
 
-	// Even on error we sent the reply so it needs to be
-	// read
 	if err := s.dec.Decode(call.Reply); err != nil && err != io.EOF {
-		call.setError(newClientError(err))
+		call.doneWithError(newClientError(err))
 		return err
 	}
+	call.done()
 	return nil
 }
+
+// makeSteram performs a streaming call, either local or remote.
+func (c *Client) makeStream(call *Call) {
+	logger.Debugf("stream: %s", call.SvcID)
+
+	// Handle local RPC calls
+	if call.Dest == "" || c.host == nil || call.Dest == c.host.ID() {
+		logger.Debugf("local call: %s", call.SvcID)
+		if c.server == nil {
+			err := &clientError{"Cannot make local calls: server not set"}
+			call.doneWithError(err)
+			return
+		}
+		err := c.server.serverStream(call)
+		call.doneWithError(err)
+		return
+	}
+
+	// Handle remote RPC calls
+	if c.host == nil {
+		panic("no host set: cannot perform remote call")
+	}
+	if c.protocol == "" {
+		panic("no protocol set: cannot perform remote call")
+	}
+	c.stream(call)
+}
+
+// stream makes a REMOTE RPC streaming call by initiating a libp2p stream to
+// the destination, writing argument channel objects to it and reading the
+// replies to the replies channel.
+func (c *Client) stream(call *Call) {
+	logger.Debug("streaming remote call")
+
+	// A helper to drain the send channel on errors.
+	drainStreamArgs := func() {
+		for {
+			_, ok := call.StreamArgs.Recv()
+			if !ok {
+				return
+			}
+		}
+	}
+
+	s, err := c.host.NewStream(call.ctx, call.Dest, c.protocol)
+	if err != nil {
+		call.doneWithError(newClientError(err))
+		drainStreamArgs()
```

Comment:
should we drain on a goroutine? We should probably not block returning on an error because the send channel is not closed. Draining can happen on a background task.

Also, what is the reason of draining. Is it just for MultiStream, because it may cause dead-locking when channels have no more space? Should this be addressed with a context? yes but it is simpler to do it this way because we would have to add a reflect case on a context that is cancelled way below.

I guess clients should be careful to stop sending on channels if the call has failed (i.e. by cancelling a context and closing).